File: frontend/src/utils/supertrendTradingView.js
/**
 * Supertrend（TradingView ta.supertrend 对齐）
 * 与 trade/market/supertrend_tradingview.py、trade-mcp KlineIndicatorCalculator 一致。
 * 默认 ATR 周期 10、乘数 3。
 */

function atrTradingView (highs, lows, closes, period) {
  const n = closes.length
  const atr = new Array(n).fill(NaN)
  if (n < period) return atr
  const trList = []
  let rma = NaN
  for (let i = 0; i < n; i++) {
    const prevClose = i > 0 ? closes[i - 1] : closes[i]
    const tr1 = highs[i] - lows[i]
    const tr2 = Math.abs(highs[i] - prevClose)
    const tr3 = Math.abs(lows[i] - prevClose)
    const tr = Math.max(tr1, tr2, tr3)
    trList.push(tr)
    if (i >= period - 1) {
      if (i === period - 1) {
        let trSum = 0
        for (let t = 0; t < period; t++) trSum += trList[t]
        rma = trSum / period
        atr[i] = rma
      } else {
        rma = (rma * (period - 1) + tr) / period
        atr[i] = rma
      }
    }
  }
  return atr
}

/**
 * @param {number[]} highs
 * @param {number[]} lows
 * @param {number[]} closes
 * @param {number} [atrPeriod=10]
 * @param {number} [multiplier=3]
 * @returns {{ line: number[], trend: number[], finalUpper: number[], finalLower: number[] }}
 */
export function computeSupertrendTradingView (highs, lows, closes, atrPeriod = 10, multiplier = 3) {
  const n = closes.length
  const atr = atrTradingView(highs, lows, closes, atrPeriod)
  const line = new Array(n).fill(NaN)
  const trend = new Array(n).fill(NaN)
  const finalUpper = new Array(n).fill(NaN)
  const finalLower = new Array(n).fill(NaN)

  for (let i = 0; i < n; i++) {
    if (Number.isNaN(atr[i])) continue

    const hl2 = (highs[i] + lows[i]) / 2
    const ub = hl2 + multiplier * atr[i]
    const lb = hl2 - multiplier * atr[i]

    if (i === 0) {
      finalUpper[i] = ub
      finalLower[i] = lb
      trend[i] = closes[i] > finalUpper[i] ? 1 : -1
      line[i] = trend[i] === 1 ? finalLower[i] : finalUpper[i]
      continue
    }

    const fuPrev = finalUpper[i - 1]
    const flPrev = finalLower[i - 1]

    if (!Number.isNaN(lb) && !Number.isNaN(flPrev)) {
      finalLower[i] = (lb > flPrev || closes[i - 1] <= flPrev) ? lb : flPrev
    } else {
      finalLower[i] = lb
    }

    if (!Number.isNaN(ub) && !Number.isNaN(fuPrev)) {
      finalUpper[i] = (ub < fuPrev || closes[i - 1] >= fuPrev) ? ub : fuPrev
    } else {
      finalUpper[i] = ub
    }

    if (trend[i - 1] === 1 && closes[i] < finalLower[i]) {
      trend[i] = -1
    } else if (trend[i - 1] === -1 && closes[i] > finalUpper[i]) {
      trend[i] = 1
    } else {
      trend[i] = trend[i - 1]
    }

    line[i] = trend[i] === 1 ? finalLower[i] : finalUpper[i]
  }

  return { line, trend, finalUpper, finalLower }
}

File: trade/market/supertrend_tradingview.py
"""
Supertrend（TradingView ta.supertrend 对齐）

默认参数：ATR 周期 10、乘数 3（与 Pine `ta.supertrend(3, 10)` 一致）。
基础价 HL2 = (high + low) / 2；ATR 为 Wilder RMA（与 `_calculate_atr_tradingview` 相同）。

上轨/下轨为 HL2 ± multiplier×ATR，再经「只单向修正」的 trailing 得到 final_upper / final_lower；
趋势翻转：上涨趋势中收盘价跌破 final_lower → 空；下跌趋势中收盘价突破 final_upper → 多。
输出线：趋势为多(1)时取 final_lower，为空(-1)时取 final_upper。
"""
from __future__ import annotations

from typing import Tuple

import numpy as np


def calculate_supertrend_tradingview(
    high: np.ndarray,
    low: np.ndarray,
    close: np.ndarray,
    atr: np.ndarray,
    multiplier: float = 3.0,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Args:
        high, low, close: 同长度
        atr: 与 ATR(period) 输出同长度，前段可为 nan
        multiplier: 通常为 3.0

    Returns:
        (line, trend, final_upper, final_lower, upper_basic)
        trend: 1.0 多头, -1.0 空头
    """
    n = len(close)
    hl2 = (high + low) * 0.5
    upper_basic = hl2 + multiplier * atr
    lower_basic = hl2 - multiplier * atr

    final_upper = np.full(n, np.nan, dtype=np.float64)
    final_lower = np.full(n, np.nan, dtype=np.float64)
    trend = np.zeros(n, dtype=np.float64)
    line = np.full(n, np.nan, dtype=np.float64)

    for i in range(n):
        if np.isnan(atr[i]):
            continue

        ub = upper_basic[i]
        lb = lower_basic[i]
        if i == 0:
            final_upper[i] = ub
            final_lower[i] = lb
            trend[i] = 1.0 if close[i] > final_upper[i] else -1.0
            line[i] = final_lower[i] if trend[i] == 1.0 else final_upper[i]
            continue

        fu_prev = final_upper[i - 1]
        fl_prev = final_lower[i - 1]

        # Trailing：下轨仅在 raw 更高或上一根收盘跌破旧下轨时取 raw；否则保持
        if not np.isnan(lb) and not np.isnan(fl_prev):
            if lb > fl_prev or close[i - 1] <= fl_prev:
                final_lower[i] = lb
            else:
                final_lower[i] = fl_prev
        else:
            final_lower[i] = lb

        if not np.isnan(ub) and not np.isnan(fu_prev):
            if ub < fu_prev or close[i - 1] >= fu_prev:
                final_upper[i] = ub
            else:
                final_upper[i] = fu_prev
        else:
            final_upper[i] = ub

        if trend[i - 1] == 1.0 and close[i] < final_lower[i]:
            trend[i] = -1.0
        elif trend[i - 1] == -1.0 and close[i] > final_upper[i]:
            trend[i] = 1.0
        else:
            trend[i] = trend[i - 1]

        line[i] = final_lower[i] if trend[i] == 1.0 else final_upper[i]

    return line, trend, final_upper, final_lower, upper_basic
